lib/ddtrace/profiling/recorder.rb

```diff
@@ -0,0 +1,23 @@
+require 'ddtrace/profiling/buffer'
+
+module Datadog
+  module Profiling
+    # Profiling buffer that stores profiling events. The buffer has a maximum size and when
+    # the buffer is full, a random event is discarded. This class is thread-safe.
+    class Recorder
+      def initialize(max_size)
+        @buffers = Hash.new do |hash, key|
+          hash[key] = Profiling::Buffer.new(max_size)
+        end
+      end
+
+      def push(event)
```

Comment:
Will the same `Recorder` be called from multiple threads? If so, we'll need to protect the lazy initialization of `Profiling::Buffer` objects.

Reply:
Yes it will so I guess your comment apply.

Reply:
So this is actually a design clarification I wanted to make...

There's a concern in my mind that the `Recorder` could end up with many buffers if its called with many different classes: shouldn't happen but could if we're not careful about what we push to it.

What do you think about being explicit on initialization? e.g. `Recorder.new([CPUEvent, MemoryEvent])` to eager create the buffers, and raise an error if it receives an unexpected event type? I think this might be safer for memory use, and would help us avoid the synchronization issue too.

Reply:
Yes, you can definitely do that.

The `Profiler` class will be the one creating both the `Recorder` and the `Collector`s so it can know what'll be recorded.